File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
       
        
        Node* iter = head;
        Node* front = head;
        
        // step 1: create copies and attach
        
        
        while(iter!=NULL){
            front = iter->next;
            
            Node* n = new Node(iter->val);
            
            iter->next = n;
            n->next = front;
            iter = front;
        }
        
        
        // step 2: assign random pointer to copy
        
        
        iter = head;
        
        while(iter!=NULL){
            
            if(iter->random!=NULL){
                iter->next->random = iter->random->next;
                
            }
            
            iter = iter->next->next;
        }
        
        
        // step 3: detach ll
        
        iter = head;
        Node* dummy = new Node(0);
        Node* temp = dummy;
        
        while(iter!=NULL){
            
            front = iter->next->next;
            
            temp->next = iter->next;
            
            iter->next = front;
            
            temp= temp->next;
            
            iter = front;
            
        }
        
        return dummy->next;
        
    }
};

File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        
        curr = head
        nxt = head
        
        while(curr):
            nxt = curr.next
            newnode = Node(curr.val,None,None)
            curr.next = newnode
            newnode.next = nxt
            curr = nxt
        
        
        curr = head
        
        while(curr):
            if(curr.random is not None):
                curr.next.random = curr.random.next
            curr = curr.next.next
        
        dummy = ListNode(-1)
        temp = dummy
        curr = head
        
        
        while(curr):
            nxt = curr.next.next
            temp.next = curr.next
            curr.next = nxt
            temp = temp.next
            curr = nxt
        
        return dummy.next
            
            
        
        
            
        
        
        